### DISCUSSION TITLE
[eframe] Custom event loop proxy?

### DISCUSSION THREAD

**Discussion by Dicklessgreat** [emilk/egui#2679] [Q&A]
Can I use main thread not only for GUI repaint but also for other things with eframe?
In my head, I want something like...
```
loop {
    match message_manager.get_message_if_there_is_one() {
        Message::Repaint => { app.update(ctx, frame); },
        Message::AutoSave => { my_model.save_state(); },
        ...
    }
}
```
and other part of my module, let's say `timer` crate.
```
fn auto_save(&self) {
    if self.check_time() {
        message_manager.post_message(Message::AutoSave);
    }
}
```
It seems like `Context::request_repaint()` can post `Message::Repaint`, but can I invoke other tasks to the main thread?
Or, should I implement custom eframe-like crate?

**Comment by imxood** [comment]
I also encountered the same problem, your idea is very good, but don’t know how to achieve it...

**Reply by Dicklessgreat** [reply]
At first I did something like the title, but now I create a separate thread and let it manage the application status.
It is difficult to exchange messages between threads, but I think it is not a bad choice.

By the way, I have tried to implement the title case.
I don't remember how it looked like, since I've already thrown away the source code, but it was to add a message()-like method to the eframe::App trait, and the user could implement that method freely and use the main thread.
I added another enumerator like MessageReceived to eframe::UserEvent and made a match arm for that event to call message() as mentioned earlier.
However, after saying so myself, I'm not sure if the egui side should do that or not. I think it would be nice to have something like that as a simple implementation.

**Comment by emilk** [comment]
I think this could be implemented in `eframe`. I created an issue for it:
* https://github.com/emilk/egui/issues/3256
